content/frontend-web/react-typescript/migrate-class-component-to-functional-component.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Migrate Class Component to Functional Component
+showMetadata: false
+editable: true
+showToc: true
+order: 3
+---
+
+# Local State
+
+เราสามารถแปลงจาก [Class Component](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/class_components/) ด้านล่างนี้
+
+```ts
+import React from "react";
+
+interface MyProps {
+  message: string;
+};
+interface MyState {
+  count: number; 
+};
+class App extends React.Component<MyProps, MyState> {
+  state: MyState = {
+    // optional second annotation for better type inference
+    count: 0,
+  };
+  render() {
+    return (
+      <div>
+        {this.props.message} {this.state.count}
+      </div>
+    );
+  }
+}
+```
+
+ให้เป็น [Functional Component](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components) แบบนี้ได้ 
+
+```ts
+import React, { useState } from "react";
+
+interface PropTypes {
+  message: string;
+}
+
+const App = ({ message }: PropTypes) => {
+  const [count, setCount] = useState(0);
+
+  return (
+    <div>
+      {message} {count}
+    </div>
+  );
+};
+export default App;
+```
+
+ทำไมถึงไม่แนะนำ `React.FC` หรือ `React.FunctionComponent`/`React.VoidFunctionComponent` อ่านรายละเอียดใน [Functional Component](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components) 
+
+# Handle Click Event
+
+เราสามารถแปลงจาก [Class Component](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/class_components) ด้านล่างนี้
+
+```ts
+import React from "react";
+
+interface PropTypes {
+  message: string;
+}
+interface StateTypes {
+  counter: number;
+}
+class App extends React.Component<PropTypes, StateTypes> {
+  state: StateTypes = {
+    counter: 0
+  };
+
+  handleClick(amount: number) {
+    this.setState( state => ({
+      counter: state.counter + amount
+    }));
+  }
+
+  render() {
+    return (
+      <div>
+        <h1>{this.props.message}</h1>
+        <button onClick={() => this.handleClick(1)}> Increase </button>
+        <p>{this.state.counter}</p>
+      </div>
+    );
+  }
+}
+
+export default App;
+```
+
+ให้เป็น Functional Component แบบนี้ได้ 
+
+```ts
+import React, { useState } from "react";
+
+interface PropTypes {
+  message: string;
+}
+
+const App = ({ message }: PropTypes) => {
+  const [counter, setCounter] = useState(0);
+
+  const handleClick = (amount: number) => {
+    setCounter(counter + amount);
+  };
+
+  return (
+    <div>
+      <h1>{message}</h1>
+      <button onClick={() => handleClick(1)}> Increase </button>

```

Comment:
Could you please check if we should avoid use arrow function in event handler because it can impact a performance because it will be processed whenever a component re-render from a changed state?

I think we we should use ` onClick={handleClick(1)}` instead. 


You can also show a curry function handler that we can accept `event`. For example:
```ts
const handleClick = (amount: number) => (event: : React.MouseEvent<HTMLElement>) => {
    setCounter(counter + amount);
};
```

Reply:
Good, this way is more convenient, but it still consumes some additional create function when re-render process as shown in the [React official docs](https://reactjs.org/docs/faq-functions.html#arrow-function-in-render). I'll add some note for the reader.

Reply:
@mildronize  Could you please check if you can change click event to `<button onClick={handleClick(1)}> Increase </button>`again? I think it should work with a function component.

Reply:
@aaronamm the ` onClick={handleClick(1)}`  will not be work both. it will cause an infinite loop of rerendering ui, as you can see in the figure.

![image](https://user-images.githubusercontent.com/3647850/124539861-73fe7680-de48-11eb-8694-a3a8d38492f7.png)

Reply:
@mildronize, Interesting, I will merge your PR and I will take a look later.

Thanks.

Reply:
@aaronamm it's also causing problems in the functional components. 
![image](https://user-images.githubusercontent.com/3647850/124540661-179c5680-de4a-11eb-82ed-e077ec008d54.png)